### PR TITLE
Reload config update pinned_images

### LINKED ISSUE #7954: crio v1.29.2 config reload does not update pinned_images
### What happened?

```
$ crio --version
crio version 1.29.2-9.rhaos4.16.git5052e1a.el9
Version:        1.29.2-9.rhaos4.16.git5052e1a.el9
```
If I update pinned_images  config

```
[crio]
  [crio.image]
    pinned_images = ["quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:86d26e7ebcccd6f07a75db5b1e56283b25c2ee1c6a755d6ffc5a4d59beb9c504"]
```
reload service
```
$ systemctl reload crio.service
```
config appearss to reload correct after catching HUP signal.
```
$ journalctl -u crio | grep "pinned images"
Apr 02 17:44:09 ip-10-0-14-61 crio[1096516]: time="2024-04-02 17:44:09.569932578Z" level=info msg="Updated new pinned images: [quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:86d26e7ebcccd6f07a75db5b1e56283b25c2ee1c6a755d6ffc5a4d59beb9c504]
```

client reports the image as unpinned
```
crictl image --pinned quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:86d26e7ebcccd6f07a75db5b1e56283b25c2ee1c6a755d6ffc5a4d59beb9c504
IMAGE                                            TAG                 IMAGE ID            SIZE                PINNED
quay.io/openshift-release-dev/ocp-v4.0-art-dev   <none>              7948809a7f475       672MB               false
```
if I restart the service it shows pinned

### What did you expect to happen?

reload of config would update pinned_images

### How can we reproduce it (as minimally and precisely as possible)?

this was tested on nightly 4.16 payload

### Anything else we need to know?

_No response_

### CRI-O and Kubernetes version

```
$ crio --version
crio version 1.29.2-9.rhaos4.16.git5052e1a.el9
Version:        1.29.2-9.rhaos4.16.git5052e1a.el9
```

kubelet --version
Kubernetes v1.29.2+258f1d5

### OS version

```
Linux ip-10-0-14-61 5.14.0-432.el9.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Mar 20 20:26:59 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```
```
# cat /etc/os-release
NAME="CentOS Stream CoreOS"
ID="rhcos"
ID_LIKE="rhel fedora"
VERSION="416.94.202403300731-0"
VERSION_ID="4.16"
VARIANT="CoreOS"
VARIANT_ID=coreos
PLATFORM_ID="platform:el9"
PRETTY_NAME="CentOS Stream CoreOS 416.94.202403300731-0"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:9::coreos"
HOME_URL="https://centos.org/"
DOCUMENTATION_URL="https://docs.okd.io/latest/welcome/index.html"
BUG_REPORT_URL="https://access.redhat.com/labs/rhir/"
REDHAT_BUGZILLA_PRODUCT="OpenShift Container Platform"
REDHAT_BUGZILLA_PRODUCT_VERSION="4.16"
REDHAT_SUPPORT_PRODUCT="OpenShift Container Platform"
REDHAT_SUPPORT_PRODUCT_VERSION="4.16"
OPENSHIFT_VERSION="4.16"
RHEL_VERSION=9.4
OSTREE_VERSION="416.94.202403300731-0"
```

### Additional environment details (AWS, VirtualBox, physical, etc.)

aws OCP 4.16 nightly

### PR DESCRIPTION
**Pull request by roman-kiselenko**
#### What type of PR is this?

/kind bug
/kind feature

#### What this PR does / why we need it:

Updates pinned images list on config reload

#### Which issue(s) this PR fixes:

Fixes #7954 

#### Does this PR introduce a user-facing change?

None

```release-note

Updates pinned images list on config reload

```

/cc @haircommander @sohankunkerkar @hexfusion @kwilczynski

### CODE REVIEW

File: server/server.go
Comment (sohankunkerkar) [inline server/server.go:591]: I think instead of checking the length, how about we change the logic of `ReloadPinnedImages` to return a boolean type? This will ensure that `UpdateImageCache` is called only when `ReloadPinnedImages` returns true indicating an update to the list of pinned_images.

**Review by sohankunkerkar — COMMENTED** [review]
@roman-kiselenko Thanks for working on this issue. There's already an integration test for this to cover the use case of updating the pinned_images when cri-o restarts, could you add another one when cri-o reloads?

**Comment by sohankunkerkar** [comment]
How about we update the image cache in `reloadConfigByHUP` only when there's an update to the list of pinned_images, instead of passing the `updateImageCache func([]string)` to `Reload`?
```go
func (c *Config) ReloadPinnedImages(newConfig *Config) bool {
	updatedPinnedImages := make([]string, len(newConfig.PinnedImages))
	updatedPinnedImageList := false

	for i, image := range newConfig.PinnedImages {
		if i < len(c.PinnedImages) && image == c.PinnedImages[i] {
			continue
		}
		updatedPinnedImages[i] = image
		updatedPinnedImageList = true
	}

	if updatedPinnedImageList {
		logrus.Infof("Updated new pinned images: %+v", updatedPinnedImages)
		c.PinnedImages = updatedPinnedImages
	}
	
	return updatedPinnedImageList
}
```

And then in reloadConfigByHUP, use this:
```go
// reloadConfigByHUP reload the config and UpdateImageCache
func (s *Server) reloadConfigByHUP() error {
	if err := s.config.Reload(); err != nil {
		return err
	}
	if if s.config.ReloadPinnedImages() {
		if err := s.StorageImageServer().UpdateImageCache(s.config.PinnedImages); err != nil {
			return err
		}
	}
	return nil
}
```

**Comment by roman-kiselenko** [comment]
> How about we update the image cache in `reloadConfigByHUP` only when there's an update to the list of pinned_images, instead of passing the `updateImageCache func([]string)` to `Reload`?
> 
> ```go
> func (c *Config) ReloadPinnedImages(newConfig *Config) bool {
> 	updatedPinnedImages := make([]string, len(newConfig.PinnedImages))
> 	updatedPinnedImageList := false
> 
> 	for i, image := range newConfig.PinnedImages {
> 		if i < len(c.PinnedImages) && image == c.PinnedImages[i] {
> 			continue
> 		}
> 		updatedPinnedImages[i] = image
> 		updatedPinnedImageList = true
> 	}
> 
> 	if updatedPinnedImageList {
> 		logrus.Infof("Updated new pinned images: %+v", updatedPinnedImages)
> 		c.PinnedImages = updatedPinnedImages
> 	}
> 	
> 	return updatedPinnedImageList
> }
> ```
> 
> And then in reloadConfigByHUP, use this:
> 
> ```go
> // reloadConfigByHUP reload the config and UpdateImageCache
> func (s *Server) reloadConfigByHUP() error {
> 	if err := s.config.Reload(); err != nil {
> 		return err
> 	}
> 	if if s.config.ReloadPinnedImages() {
> 		if err := s.StorageImageServer().UpdateImageCache(s.config.PinnedImages); err != nil {
> 			return err
> 		}
> 	}
> 	return nil
> }
> ```

cool, I'm going to use your solutions, thanks 🙏

**Comment by sohankunkerkar** [comment]
> The only thing here is how to pass `newConfig` variable. That the reason for passing `updateImageCache func([]string)` to `Reload()` 🤔
> 
> ```go
> s.config.ReloadPinnedImages() 
> ```


hmm... I think you need to wrap `ReloadPinnedImages` in another function, maybe?

```go
func (c *Config) IsPinnedImagesUpdated() bool {
	return c.ReloadPinnedImages(c)
}
```

File: pkg/config/reload.go
Comment (haircommander) [inline pkg/config/reload.go:17]: is it possible to use the `_inject_test.go` pattern where we can override for the tests specifically by calling a hidden setter?

Comment (roman-kiselenko) [reply pkg/config/reload.go]: I've changed a bit of implementation and used another way of passing `Config.` PTAL

**Comment by roman-kiselenko** [comment]
> > The only thing here is how to pass `newConfig` variable. That the reason for passing `updateImageCache func([]string)` to `Reload()` 🤔
> > ```go
> > s.config.ReloadPinnedImages() 
> > ```
> 
> hmm... I think you need to wrap `ReloadPinnedImages` in another function, maybe?
> 
> ```go
> func (c *Config) IsPinnedImagesUpdated() bool {
> 	return c.ReloadPinnedImages(c)
> }
> ```

I've changed a bit of implementation and used another way of passing Config. PTAL

**Comment by roman-kiselenko** [comment]
> > The only thing here is how to pass `newConfig` variable. That the reason for passing `updateImageCache func([]string)` to `Reload()` 🤔
> > ```go
> > s.config.ReloadPinnedImages() 
> > ```
> 
> hmm... I think you need to wrap `ReloadPinnedImages` in another function, maybe?
> 
> ```go
> func (c *Config) IsPinnedImagesUpdated() bool {
> 	return c.ReloadPinnedImages(c)
> }
> ```

The `Reload()` function contains a `newConfig` variable with fresh PinnedImages. I will rewrite the `Reload()` logic a little so the `PinnedImages` attribute will update appropriately.

**Review by sohankunkerkar — APPROVED** [review]
Changes LGTM
/lgtm

**Comment by sohankunkerkar** [comment]
@cri-o/cri-o-maintainers PTAL

File: internal/storage/image.go
Comment (haircommander) [inline internal/storage/image.go:904]: nit: invalidates*

File: test/reload_config.bats
Comment (haircommander) [inline test/reload_config.bats:241]: I think this test could do a `crictl images -o json` and find the pinned images, then do so after we do the reload and make sure the images are updated

File: server/server.go
Comment (haircommander) [inline server/server.go:581]: It's not clear to me how this condition will be true if we're replacing the config in place in Reload(). won't newConfig == s.config at this point?

Comment (roman-kiselenko) [reply server/server.go:581]: The `Reload()` does not replace the config; it only updates the `newConfig` variable, and all the changes vanish with returns.

https://github.com/cri-o/cri-o/blob/0ae4367ef3b337894b13cd199b007653a5c2b6be/pkg/config/reload.go#L26-L38

File: test/reload_config.bats
Comment (roman-kiselenko) [reply test/reload_config.bats:241]: This comment led me to understand that the current implementation is wrong, and it doesn't update the pinned attribute :) 

Thanks 🙏

Comment (haircommander) [reply test/reload_config.bats:241]: I trust the config hadn't pinned anything else, but just to future proof, can you also do the opposite check before we reload_crio?

File: server/server.go
Comment (haircommander) [reply server/server.go]: wait but we do replace the c.config in reload: https://github.com/cri-o/cri-o/blob/0ae4367ef3b337894b13cd199b007653a5c2b6be/pkg/config/reload.go#L148-L156 we absorb the newConfig, and then take select changes into the c.config (old config) so how would the newConfig and c.config ever differ  WRT pinned images?

Comment (roman-kiselenko) [reply server/server.go]: We're merging the old config pinned image list with the new pinned image list, and if there are changes, we update the pinned image list in `imageService`. Because we're relying on the returned state from `ReloadPinnedImages` we can perform `imageService` update only outside the Reload method.

**Comment by hexfusion** [comment]
Just checking in on this how are we feeling about the current direction @haircommander @sohankunkerkar

File: internal/storage/image.go
Comment (haircommander) [inline internal/storage/image.go:905]: I'm not sure we should append here. Wouldnt' we just want to overwrite the slice? someone could theoretically unpin an image and we'd want it to qualify for GC then

File: pkg/config/reload.go
Comment (hexfusion) [inline pkg/config/reload.go:17]: this feels a little awkward since Reload is a method of Config.

File: pkg/config/reload_test.go
Comment (hexfusion) [inline pkg/config/reload_test.go:38]: In this use case sut is aleady reference to `Config` right?

**Comment by roman-kiselenko** [comment]
@haircommander @sohankunkerkar @hexfusion 

The trick with `Reload()` and return config confuses readers. I've changed the implementation to make it more readable. 

The `Reload()` method now updates the config struct, and fields `PinnedImages` and `PauseImage` are passed to `UpdatePinnedImagesList`.
https://github.com/cri-o/cri-o/blob/fbb899bb89a73917da0649a5c702964637454b32/internal/storage/image.go#L857-L859 

Both fields are updated in ImageService; hence, we must update both when HUP happens.

Also, I've added a test to catch some corner cases.

@cri-o/cri-o-maintainers PTAL

**Comment by haircommander** [comment]
Yes I like this better! Good rewrite, and thanks for your patience
/approve
Tagging another @cri-o/cri-o-maintainers for the final review

File: server/server.go
Comment (haircommander) [inline server/server.go:583]: Totally non blocking nit: I think i'd prefer combining the config fields here instead of passing two args

File: internal/storage/image.go
Comment (haircommander) [inline internal/storage/image.go:906]: If pause image is changed to ", what happens here? I forget, do we fallback to anything in sandbox_run_linux.go, and if so should we include it here?

**Review by haircommander — APPROVED** [review]
Actually a couple of notes but non blocking

File: internal/storage/image.go
Comment (roman-kiselenko) [reply internal/storage/image.go:906]: That's a good catch; I'll update it. :pray:

Comment (roman-kiselenko) [reply internal/storage/image.go:906]: I'm a bit confused; could you elaborate? 

AFAIU, the pause image in runPodSandbox comes from the config. Hence, it will be updated on HUP.  

https://github.com/cri-o/cri-o/blob/fbb899bb89a73917da0649a5c702964637454b32/server/sandbox_run_linux.go#L461-L464

Comment (roman-kiselenko) [reply internal/storage/image.go]: > If pause image is changed to "

This will not happen because the empty value in the config will not be parsed

> Unable to reload configuration: invalid reference format

Hence, `DefaultPauseImage` will always be in the list of `PinnedImages` and as `PauseImage`.

File: pkg/config/reload.go
Comment (sohankunkerkar) [inline pkg/config/reload.go:78]: why do we have two entries here?

Comment (roman-kiselenko) [reply pkg/config/reload.go:78]: ah, bad rebase thx

File: pkg/config/reload.go
Comment (hexfusion) [inline pkg/config/reload.go:148]: optional: Might make sense to add a few words on what the bool this public method now returns indicates?

File: server/server.go
Comment (sohankunkerkar) [inline server/server.go:582]: With this change, you can drop this code: https://github.com/cri-o/cri-o/blob/main/internal/storage/image.go#L857

File: pkg/config/reload.go
Comment (hexfusion) [inline pkg/config/reload.go:147]: :+1:

File: internal/storage/image.go
Comment (sohankunkerkar) [reply internal/storage/image.go]: Could you add an integration test for this?

Comment (sohankunkerkar) [reply internal/storage/image.go]: Just to make sure everything works as expected.

File: pkg/config/reload.go
Comment (roman-kiselenko) [reply pkg/config/reload.go:146]: Thx, I've rewritten this method and removed the `bool` return. Not log print happens in the same way as in other methods with usage of `logConfig()`

File: internal/storage/image.go
Comment (roman-kiselenko) [reply internal/storage/image.go]: I've implemented that checking here https://github.com/cri-o/cri-o/pull/7976/files#diff-4ea762d405a8752ac8b728bfbe94b8ee5f962013a1e65d4e5d20208556418447R255-R257

File: server/server.go
Comment (sohankunkerkar) [reply server/server.go:582]: ah, right!

**Comment by roman-kiselenko** [comment]
The code has been changed a lot since last approve, I've re-request review 🙏

**Review by hexfusion — COMMENTED** [review]
I am not a maintainer of this code but this approach looks much better :+1:

**Comment by haircommander** [comment]
/approve
Still, @sohankunkerkar for the final lgtm?

**Comment by hexfusion** [comment]
The integrations might be flakes but I think you need to fix the shfmt linter
```
make shfmt
```
probably looking for whitespace

**Comment by sohankunkerkar** [comment]
> The integrations might be flakes but I think you need to fix the shfmt linter
> 
> ```
> make shfmt
> ```
> 
> probably looking for whitespace

@roman-kiselenko Thanks for your patience. One last fix before `/lgtm`

**Comment by hexfusion** [comment]
fedora-kata test is failing due to rate limiting
> rateLimitExceeded

**Comment by roman-kiselenko** [comment]
> fedora-kata test is failing due to rate limiting
> 
> > rateLimitExceeded

[integration / userns / amd64](https://github.com/cri-o/cri-o/actions/runs/8737513570/job/23974871352?pr=7976#step:13:416) - failed and I cant find a reason. 🤔 

[integration / conmon / arm64](https://github.com/cri-o/cri-o/actions/runs/8737513570/job/23974870432?pr=7976#step:13:352) - NRI tests are flaky

**Review by sohankunkerkar — APPROVED** [review]
/lgtm

**Comment by sohankunkerkar** [comment]
>last failure might not of been a flake

Yup, it's a legit one.

**Comment by hexfusion** [comment]
cc @roman-kiselenko PTAL

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/cri-o_cri-o/7976/pull-ci-cri-o-cri-o-main-ci-fedora-kata/1780966933546405888/artifacts/fedora-kata/cri-o-gather/artifacts/testout.txt
```
not ok 216 reload config should update 'pinned_images' and 'pause_image' is pinned automatically
# (in test file reload_config.bats, line 260)
#   `[[ "$output" == *"wasm"* ]]' failed
```

**Comment by roman-kiselenko** [comment]
> cc @roman-kiselenko PTAL
> 
> https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/cri-o_cri-o/7976/pull-ci-cri-o-cri-o-main-ci-fedora-kata/1780966933546405888/artifacts/fedora-kata/cri-o-gather/artifacts/testout.txt
> ```
> not ok 216 reload config should update 'pinned_images' and 'pause_image' is pinned automatically
> # (in test file reload_config.bats, line 260)
> #   `[[ "$output" == *"wasm"* ]]' failed
> ```

I can't reproduce it locally :sad:

going to poke around logs for a while

**Comment by roman-kiselenko** [comment]
> I think this failure is only for Kata; the other two integration tests are looking fine

Do you think we can disable them for Kata?

I can't find the reason why it failed.

**Comment by sohankunkerkar** [comment]
@littlejawa  Do you have any idea here?

**Comment by sohankunkerkar** [comment]
```sh
 level=error msg="Getting /opt/kata/bin/containerd-shim-kata-v2 OCI runtime features failed: io.containerd.kata.v2: shim namespace cannot be empty: exit status 1" file="config/config.go:1306"
```

**Comment by roman-kiselenko** [comment]
> ```shell
>  level=error msg="Getting /opt/kata/bin/containerd-shim-kata-v2 OCI runtime features failed: io.containerd.kata.v2: shim namespace cannot be empty: exit status 1" file="config/config.go:1306"
> ```

That looks like a problem; I've found this https://github.com/kata-containers/kata-containers/issues/541 

TBH I don't know how CRI-O works with Kata.  😞

**Comment by roman-kiselenko** [comment]
Also [image pull & list disable on Kata](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/cri-o_cri-o/8019/pull-ci-cri-o-cri-o-main-ci-fedora-kata/1780041107673976832/artifacts/fedora-kata/cri-o-gather/artifacts/testout.txt#:~:text=ok%2030%20image%20pull%20and%20list%20by%20manifest,and%20manifest%20list%20digest%27%20not%20working%20in%20kata)

File: contrib/test/ci/vars.yml
Comment (sohankunkerkar) [inline contrib/test/ci/vars.yml:47]: There's already a [PR](https://github.com/cri-o/cri-o/pull/7958) to test the new version of Kata.

Comment (roman-kiselenko) [reply contrib/test/ci/vars.yml:47]: oh, thanks, I didn't spot this.

**Comment by sohankunkerkar** [comment]
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/cri-o_cri-o/7976/pull-ci-cri-o-cri-o-main-ci-fedora-integration/1782419051419537408/artifacts/fedora-integration/cri-o-gather/artifacts/testout.txt

I think [this](https://github.com/cri-o/cri-o/pull/8057) should solve the issue.

**Comment by hexfusion** [comment]
Have we tested inside a kata instance to verify what is going on? Does the test provide the details of how to do this?

**Comment by hexfusion** [comment]
cc @roman-kiselenko

**Comment by roman-kiselenko** [comment]
> We can skip this test here: https://github.com/cri-o/cri-o/blob/main/contrib/test/ci/vars.yml#L101 and open an issue to track the test failure.

I've skipped this test for `kata.` I'm going to create an issue for this after a successful merge.

**Review by sohankunkerkar — APPROVED** [review]
/lgtm

**Comment by hexfusion** [comment]
cc @sohankunkerkar can we get a cherry pick to 1.29?

**Comment by hexfusion** [comment]
thanks for seeing this through @roman-kiselenko

**Comment by littlejawa** [comment]
Hey there,
Sorry for being late on this one, I was offline for a while. I'll have a look at the issue and try to understand what's wrong on the kata side.

**Comment by roman-kiselenko** [comment]
> Hey there,
> Sorry for being late on this one; I was offline for a while. I'll have a look at the issue and try to understand what's wrong on the kata side.

hey, thank you for your time.

We have an umbrella issue here https://github.com/cri-o/cri-o/issues/8074

**Pull request merged**
